### PR TITLE
Ranked Farthest First Traversal

### PR DESCRIPTION
**Pull request by samuelstanton**
Efficient implementation of farthest first traversal using `heapq`. Avoids constructing the full pairwise distance matrix, so should be suitable for traversal of very large libraries. Option to pass ranking scores to choose first point and break ties

Possible areas for improvement

- [x] support arbitrary distance functions
- [x] only put ranking scores in heap if they are actually passed
- [x] input types could just be lists, undecided
- [ ] switch to weight-biased leftist tree?

### CODE REVIEW

File: src/beignet/_ranked_fft.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import heapq
+import numpy as np
+from typing import Optional
+
+
+def edit_dist(x: str, y: str):
+    """
+    Computes the edit distance between two strings.
+    """
+    return edlib.align(x, y)["editDistance"]
```
Comment (0x00b1) [inline src/beignet/_ranked_fft.py:11]: I don’t want to carry edlib unless (or until) it’s absolutely necessary. 

I believe edlib uses Myers’. I could implement it if you don’t have the time. We could also use Levenshtein.

File: src/beignet/_ranked_fft.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+
+def ranked_fft(
```
Comment (0x00b1) [inline src/beignet/_ranked_fft.py:14]: Use a more descriptive name, e.g., `ranked_farthest_first_traversal` (or even `farthest_first_traversal`).

File: src/beignet/_ranked_fft.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    library: np.ndarray,
+    ranking_scores: Optional[np.ndarray] = None,
+    n: int = 2,
+    descending: bool = False,
+):
```
Comment (0x00b1) [inline src/beignet/_ranked_fft.py:19]: Add the return type.

File: src/beignet/_ranked_fft.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+def edit_dist(x: str, y: str):
+    """
+    Computes the edit distance between two strings.
+    """
+    return edlib.align(x, y)["editDistance"]
+
+
+def ranked_fft(
+    library: np.ndarray,
+    ranking_scores: Optional[np.ndarray] = None,
```
Comment (0x00b1) [inline src/beignet/_ranked_fft.py:16]: Use the literal type syntax, i.e., numpy.ndarray | None.

File: src/beignet/_ranked_fft.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
+import edlib
+import heapq
+import numpy as np
```
Comment (0x00b1) [inline src/beignet/_ranked_fft.py:3]: No import aliases.

File: src/beignet/_ranked_fft.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
+import edlib
```
Comment (0x00b1) [inline src/beignet/_ranked_fft.py:1]: Use `isort` style (i.e., standard library imports should be separated from third-party imports).

File: src/beignet/_ranked_fft.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        remaining_indices = list(range(library.shape[0]))
+    else:
+        if descending:
+            ranking_scores = -ranking_scores
+        remaining_indices = list(np.argsort(ranking_scores))
+
+    selected = [remaining_indices.pop(0)]
+
+    if n == 1:
+        return np.array(selected)
```
Comment (0x00b1) [inline src/beignet/_ranked_fft.py:44]: Return a PyTorch `Tensor`.

File: src/beignet/_ranked_fft.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+                    edit_dist(library[idx], library[selected[i]])
+                    for i in range(num_checked, len(selected))
+                )
+                min_dist = min(min_dist, -neg_dist)
+                heapq.heappush(pq, (-min_dist, score, idx, len(selected)))
+            else:
+                selected.append(idx)
+                break
+
+    return np.array(selected)
```
Comment (0x00b1) [inline src/beignet/_ranked_fft.py:77]: Return a PyTorch `Tensor`.

File: src/beignet/_ranked_fft.py (hunk 1 above)
Comment (samuelstanton) [reply src/beignet/_ranked_fft.py:11]: yep, gotta keep dependencies under control. This is maybe another reason to change to an arbitrary distance function and let the user decide what they want to use. I used to use Levenshtein and then switched to edlib for reasons I don't remember

Comment (samuelstanton) [reply src/beignet/_ranked_fft.py]: for now I've punted by just adding a `distance_fn` arg

**Pull request merged**